components/audio_player/include/audio_player.h

```diff
@@ -0,0 +1,150 @@
+/**
+ * @file
+ * @version 0.1
+ *
+ * @copyright Copyright 2021 Espressif Systems (Shanghai) Co. Ltd.
+ * @copyright Copyright 2022 Chris Morgan <chmorgan@gmail.com>
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License");
+ *      you may not use this file except in compliance with the License.
+ *      You may obtain a copy of the License at
+ *
+ *               http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *      Unless required by applicable law or agreed to in writing, software
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
+ */
+
+/**
+ * Design notes
+ *
+ * - There is a distinct event for playing -> playing state transitions.
+ * COMPLETED_PLAYING_NEXT is helpful for users of the audio player to know
+ * the difference between playing and transitioning to another audio file
+ * vs. detecting that the audio file transitioned by looking at
+ * events indicating IDLE and then PLAYING within a short period of time.
+ *
+ * State machine diagram
+ *
+ * cb is the callback function registered with audio_player_callback_register()
+ *
+ *             cb(PLAYING)                     cb(PLAYING)
+ *   _______________________________     ____________________________________
+ *   |                             |     |                                  |
+ *   |                             |     |                                  |
+ *   |         cb(IDLE)            V     V             cb(PAUSE)            |
+ * Idle <------------------------  Playing  ----------------------------> Pause
+ *   ^                             |_____^                                  |
+ *   |                      cb(COMPLETED_PLAYING_NEXT)                      |
+ *   |                                                                      |
+ *   |______________________________________________________________________|
+ *                                cb(IDLE)
+ *
+ */
+
+#pragma once
+
+#include <stddef.h>
+#include <stdio.h>
+#include "esp_err.h"
+#include "driver/i2s.h"
```

Comment:
Just an FYI:
`driver/i2s.h` driver will become obsolete in IDF version 5.0. You can still use it, but you'll get deprecation warnings (they can be turned off). ATM, I'd stick to the old driver, because it works on all active esp-idf releases.

More information is in our [migration guide](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/migration-guides/release-5.x/peripherals.html#i2s-driver)